### PR TITLE
add dmclusters CRD

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
https://github.com/pingcap/tidb-operator/issues/2868
TODO1: dmclusters CRD

### What is changed and how does it work?
Add dmclustes CRD

### Check List 

Tests 

 - Unit test
 - E2E test
 - Stability test
 - No code

Code changes

 - Has Go code change

### Does this PR introduce a user-facing change?:

```release-note
NONE
```

### CODE REVIEW

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 1]
```diff
@@ -1437,3 +1441,272 @@ type IngressSpec struct {
...
+
+// +k8s:openapi-gen=true
+// DMClusterSpec describes the attributes that a user creates on a dm cluster
+type DMClusterSpec struct {
+	// Discovery spec
+	Discovery DMDiscoverySpec `json:"discovery,omitempty"`
+
+	// dm-master cluster spec
+	// +optional
+	Master *MasterSpec `json:"dm_master,omitempty"`
```
Comment (lonng) [inline pkg/apis/pingcap/v1alpha1/types.go:1476]: The json tag should be `master`?

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 2]
```diff
@@ -1437,3 +1441,272 @@ type IngressSpec struct {
...
+
+	// dm-worker cluster spec
+	// +optional
+	Worker *WorkerSpec `json:"dm_worker,omitempty"`
```
Comment (lonng) [inline pkg/apis/pingcap/v1alpha1/types.go:1480]: ditto?

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 1 above)
Comment (lichunzhu) [reply pkg/apis/pingcap/v1alpha1/types.go:1476]: `tiup-dm` uses `dm_master`
https://github.com/pingcap/tiup/blob/31ebd2a7bcaf32c1e0ebc87991818ab6011e83c2/components/dm/spec/topology_dm.go#L82

Comment (lonng) [reply pkg/apis/pingcap/v1alpha1/types.go:1476]: @july2993 Can we change it `master_servers/worker_servers`?

Comment (july2993) [reply pkg/apis/pingcap/v1alpha1/types.go:1476]: I think just `dm_master` is more concise?
there are no servers, just specified the master or worker's config.

like tiup-cluster.

Comment (july2993) [reply pkg/apis/pingcap/v1alpha1/types.go:1476]: currently
```
server_configs:
    dm_master:
      rpc-timeout: "30s" # you many need update test_upgrade.sh if change this cause it's used to test edit-config
      rpc-rate-limit: 10.0
      rpc-rate-burst: 40
dm_master_servers:
    - host: 172.19.0.101
```

if change
```
server_configs:
    master_servers:
      rpc-timeout: "30s" # you many need update test_upgrade.sh if change this cause it's used to test edit-config
      rpc-rate-limit: 10.0
      rpc-rate-burst: 40
dm_master_servers:
    - host: 172.19.0.101
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 3]
```diff
@@ -47,6 +47,10 @@ const (
 	TiFlashMemberType MemberType = "tiflash"
 	// TiCDCMemberType is ticdc container type
 	TiCDCMemberType MemberType = "ticdc"
+	// DMMasterMemberType is dm-master container type
+	DMMasterMemberType MemberType = "dm-master"
+	// DMWorkerMemberType is dm-worker container type
+	DMWorkerMemberType MemberType = "dm-master"
```
Comment (weekface) [inline pkg/apis/pingcap/v1alpha1/types.go:53]:
```suggestion
	DMWorkerMemberType MemberType = "dm-worker"
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 4]
```diff
@@ -1437,3 +1441,272 @@ type IngressSpec struct {
...
+
+	// +k8s:openapi-gen=false
+	// Most recently observed status of the dm cluster
+	Status DMClusterStatus `json:"status"`
+}
+
+// +k8s:openapi-gen=true
+// DMDiscoverySpec contains details of Discovery members for dm
+type DMDiscoverySpec struct {
+	corev1.ResourceRequirements `json:",inline"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1463]: Suggest not syncing Discovery component, just connect to the discovery of the tidb cluster.

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 5]
```diff
@@ -1437,3 +1441,272 @@ type IngressSpec struct {
...
+	// Host indicates the existed TiDB discovery host. If not specified, dm cluster will start a new one
+	Host string `json:"host"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1465]: Change to address?

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 6]
```diff
@@ -1437,3 +1441,272 @@ type IngressSpec struct {
...
+}
+
+// +k8s:openapi-gen=true
+// DMClusterSpec describes the attributes that a user creates on a dm cluster
+type DMClusterSpec struct {
+	// Discovery spec
+	Discovery DMDiscoverySpec `json:"discovery,omitempty"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1472]:
```suggestion
	Discovery DMDiscoverySpec `json:"discovery"`
```

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 1 above)
Comment (DanielZhangQD) [reply pkg/apis/pingcap/v1alpha1/types.go:1476]: `master` +1

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 2 above)
Comment (DanielZhangQD) [reply pkg/apis/pingcap/v1alpha1/types.go:1480]: +1

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 7]
```diff
@@ -1437,3 +1441,272 @@ type IngressSpec struct {
...
+	// +optional
+	ImagePullSecrets []corev1.LocalObjectReference `json:"imagePullSecrets,omitempty"`
+
+	// ConfigUpdateStrategy determines how the configuration change is applied to the cluster.
+	// UpdateStrategyInPlace will update the ConfigMap of configuration in-place and an extra rolling-update of the
+	// cluster component is needed to reload the configuration change.
+	// UpdateStrategyRollingUpdate will create a new ConfigMap with the new configuration and rolling-update the
+	// related components to use the new ConfigMap, that is, the new configuration will be applied automatically.
+	// +kubebuilder:validation:Enum=InPlace,RollingUpdate
+	// +kubebuilder:default=InPlacne
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1514]: Please make RollingUpdate as default, which is what we recommend for production, and for TC, the InPlace is just for backward compatibility.

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 8]
```diff
@@ -1437,3 +1441,272 @@ type IngressSpec struct {
...
+	EnablePVReclaim *bool `json:"enablePVReclaim,omitempty"`
+
+	// Whether enable the TLS connection between DM server components
+	// Optional: Defaults to nil
+	// +optional
+	TLSCluster *TLSCluster `json:"tlsCluster,omitempty"`
+
+	// Base node selectors of DM cluster Pods, components may add or override selectors upon this respectively
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1529]: May also need to add `Affinity`, `timezone`

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 9]
```diff
@@ -1437,3 +1441,272 @@ type IngressSpec struct {
...
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+}
+
+// DMClusterStatus represents the current status of a dm cluster.
+type DMClusterStatus struct {
+	Master MasterStatus `json:"dm_master,omitempty"`
+	Worker WorkerStatus `json:"dm_worker,omitempty"`
+
+	Monitor *TidbMonitorRef `json:"monitor,omitempty"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1545]: Remove this.

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 10]
```diff
@@ -1437,3 +1441,272 @@ type IngressSpec struct {
...
+// MasterSpec contains details of dm-master members
+type MasterSpec struct {
+	ComponentSpec               `json:",inline"`
+	corev1.ResourceRequirements `json:",inline"`
+
+	// The desired ready replicas
+	// +kubebuilder:validation:Minimum=1
+	Replicas int32 `json:"replicas"`
+
+	// TODO: remove optional after defaulting introduced
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1561]: Remove this

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 11]
```diff
@@ -1437,3 +1441,272 @@ type IngressSpec struct {
...
+	// Base image of the component, image tag is now allowed during validation
+	// +kubebuilder:default=pingcap/dm-master
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1563]:
```suggestion
	// +kubebuilder:default=pingcap/dm
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 12]
```diff
@@ -1437,3 +1441,272 @@ type IngressSpec struct {
...
+	// +optional
+	BaseImage string `json:"baseImage"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1565]: Add `omitempty`?

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 13]
```diff
@@ -1437,3 +1441,272 @@ type IngressSpec struct {
...
+
+	// Service defines a Kubernetes service of Master cluster.
+	// Optional: Defaults to `.spec.services` in favor of backward compatibility
+	// +optional
+	Service *ServiceSpec `json:"service,omitempty"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1570]: `ExternalTrafficPolicy` this may also need.

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 14]
```diff
@@ -1437,3 +1441,272 @@ type IngressSpec struct {
...
+// WorkerStatus is dm-worker status
+type WorkerStatus struct {
+	Synced      bool                    `json:"synced,omitempty"`
+	Phase       MemberPhase             `json:"phase,omitempty"`
+	StatefulSet *apps.StatefulSetStatus `json:"statefulSet,omitempty"`
+	Workers     map[string]WorkerMember `json:"workers,omitempty"`
+}
+
+// WorkerMember is DM Worker status
+type WorkerMember struct {
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1709]: Add a field for `health`?

File: pkg/apis/pingcap/v1alpha1/dm_config.go [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+// MasterConfig is the configuration of dm-master-server
+// +k8s:openapi-gen=true
+type MasterConfig struct {
+	// Log level.
+	// Optional: Defaults to info
+	// +optional
+	LogLevel *string `toml:"log-level" json:"log-level"`
+	// File log config.
+	// +optional
+	LogFile *string `toml:"log-file" json:"log-file"`
```
[the remark this answers is not in the document]
Comment (lichunzhu) [reply pkg/apis/pingcap/v1alpha1/dm_config.go:37]: Yes, `dm-master`/`dm-worker` will print both to stdout/stderr when `log-file` is not specified.

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 8 above)
Comment (lichunzhu) [reply pkg/apis/pingcap/v1alpha1/types.go:1529]: Why do dm need `timezone`?

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 9 above)
Comment (lichunzhu) [reply pkg/apis/pingcap/v1alpha1/types.go:1545]: Why should we remove this? How will dm set up monitor?

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 14 above)
Comment (lichunzhu) [reply pkg/apis/pingcap/v1alpha1/types.go:1709]: What is `health` used for?

File: pkg/apis/pingcap/v1alpha1/dm_config.go [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+	// +optional
+	KeepAliveTTL *int64 `toml:"keepalive-ttl,omitempty" json:"keepalive-ttl,omitempty"`
+	// dm-worker's security config
+	// +optional
+	DMSecurityConfig `toml:",inline" json:",inline"`
+}
+
+// DM common security config
+type DMSecurityConfig struct {
+	// SSLCA is the path of file that contains list of trusted SSL CAs. if set, following four settings shouldn't be empty
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/dm_config.go:85]: If this is set, the cert-allowed-cn can be optional.

File: pkg/apis/pingcap/v1alpha1/dm_config.go [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+	// +optional
+	LogLevel *string `toml:"log-level,omitempty" json:"log-level,omitempty"`
+	// File log config.
+	// +optional
+	LogFile *string `toml:"log-file,omitempty" json:"log-file,omitempty"`
+	// Log format. one of json or text.
+	// +optional
+	LogFormat *string `toml:"log-format,omitempty" json:"log-format,omitempty"`
+
+	// KeepAliveTTL is the keepalive ttl dm-worker write to dm-master embed etcd
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/dm_config.go:74]:
```suggestion
	// KeepAliveTTL is the keepalive TTL when dm-worker writes to the embedded etcd of dm-master.
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 18]
```diff
@@ -1437,3 +1448,285 @@ type IngressSpec struct {
...
+
+// +k8s:openapi-gen=true
+// DMClusterSpec describes the attributes that a user creates on a dm cluster
+type DMClusterSpec struct {
+	// Discovery spec
+	Discovery DMDiscoverySpec `json:"discovery"`
+
+	// dm-master cluster spec
+	// +optional
+	Master *MasterSpec `json:"master,omitempty"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1482]: Can `master` be empty?

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 19]
```diff
@@ -1437,3 +1448,285 @@ type IngressSpec struct {
...
+	// +optional
+	DataSubDir string `json:"dataSubDir,omitempty"`
+
+	// Config is the Configuration of dm-master-servers
+	// +optional
+	Config *MasterConfig `json:"config,omitempty"`
+}
+
+type MasterServiceSpec struct {
+	// +k8s:openapi-gen=false
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1604]: why set false here?

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 1 above)
Comment (lonng) [reply pkg/apis/pingcap/v1alpha1/types.go]: @july2993 Prefer
```yaml
server_configs:
    master:
      rpc-timeout: "30s" # you many need update test_upgrade.sh if change this cause it's used to test edit-config
      rpc-rate-limit: 10.0
      rpc-rate-burst: 40
master_servers:
    - host: 172.19.0.101
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 20]
```diff
@@ -1545,7 +1560,7 @@ type DMClusterSpec struct {
 
 // DMClusterStatus represents the current status of a dm cluster.
 type DMClusterStatus struct {
-	Master MasterStatus `json:"master,omitempty"`
+	Master MasterStatus `json:"master"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1563]:
```suggestion
	Master MasterStatus `json:"master,omitempty"`
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 21]
```diff
@@ -1479,7 +1494,7 @@ type DMClusterSpec struct {
 
 	// dm-master cluster spec
 	// +optional
-	Master *MasterSpec `json:"master,omitempty"`
+	Master *MasterSpec `json:"master"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1497]:
```suggestion
	Master MasterSpec `json:"master"`
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 22]
```diff
@@ -1498,11 +1498,11 @@ type DMClusterSpec struct {
 
 	// dm-master cluster spec
 	// +optional
-	Master *MasterSpec `json:"master"`
+	Master MasterSpec `json:"master"`
 
 	// dm-worker cluster spec
 	// +optional
-	Worker *WorkerSpec `json:"worker,omitempty"`
+	Worker WorkerSpec `json:"worker,omitempty"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1505]:
```suggestion
	Worker *WorkerSpec `json:"worker,omitempty"`
```

**Comment by lichunzhu** [comment]
@cofyc @weekface PTAL again

**Pull request merged**